### PR TITLE
cipher: add block mode traits and wrappers

### LINKED ISSUE #564: New block-modes API design
(from https://github.com/RustCrypto/block-ciphers/pull/227#issuecomment-783074558)

* At the type system, support modes requiring only an block encryption function (as opposed to both encryption and decryption)
* bring back separate types for encryptors and decryptors
* splitting the trait into two parts: core which works only on blocks (and thus padding-independent) and a buffered wrapper on top of it which would work with slices and will be generic over padding.

### PR DESCRIPTION
**Pull request by newpavlov**
Closes #564

### CODE REVIEW

**Comment by newpavlov** [comment]
Unfortunately, the block mode wrappers are a bit sub-optimal. They first copy data to an output buffer and only then encrypt/decrypt it, i.e. they effectively implement a two-pass approach. But to solve this issue we would have to expose block mode parallelism capabilities... Also ideally we would need a solution for #31. So I think it's worth to postpone it until later versions.

Another interesting observation is that mutable block cipher traits are quite symmetric with the block mode traits. I think it could be worth to explore merging them, e.g. we could say that hardware blocks implement not a block cipher, but ECB mode, on top of which we can implement other block cipher modes.

**Comment by tarcieri** [comment]
> Another interesting observation is that mutable block cipher traits are quite symmetric with the block mode traits. I think it could be worth to explore merging them, e.g. we could say that hardware blocks implement not a block cipher, but ECB mode, on top of which we can implement other block cipher modes.

Interesting idea

File: aead/src/stream.rs [hunk 1]
```diff
@@ -322,12 +323,13 @@ macro_rules! impl_stream_object {
...
-                self.stream.$in_place_op(self.position, true, associated_data, buffer)
+                self.stream
+                    .$in_place_op(self.position, true, associated_data, buffer)
             }
         }
-    }
+    };
 }
 
 impl_stream_object!(
```
Comment (tarcieri) [inline aead/src/stream.rs:335]: Huh, did `rustfmt` add new formatting rules for macros or something?

**Review by tarcieri — APPROVED** [review]

File: aead/src/stream.rs (hunk 1 above)
Comment (newpavlov) [reply aead/src/stream.rs:335]: Hm, it could be a nightly thing. I will revert the `aead` formatting changes.

**Comment by newpavlov** [comment]
The experimental changes which I currently explore have become too big, so I will merge this PR and open a new one instead.

**Pull request merged**